test/foundry/integration/UpgradeSmartAccountTest.t.sol

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "../utils/Imports.sol";
+import "../utils/NexusTest_Base.t.sol";
+
+contract UpgradeSmartAccountTest is NexusTest_Base {
+
+    function setUp() public {
+        init();
+    }
+
+    /// @notice Tests that the proxiable UUID slot is correct
+    function test_proxiableUUIDSlot() public {
+        bytes32 slot = ACCOUNT_IMPLEMENTATION.proxiableUUID();
+        assertEq(slot, 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc, "Proxiable UUID mismatch");
+    }
+
+    /// @notice Tests that the current implementation address is correct
+    function test_currentImplementationAddress() public {
+        address currentImplementation = BOB_ACCOUNT.getImplementation();
+        assertEq(currentImplementation, address(ACCOUNT_IMPLEMENTATION), "Current implementation address mismatch");
+    }
+
+    /// @notice Tests the upgrade of the smart account implementation
+    function test_upgradeImplementation() public {
+        Nexus newSmartAccount = new Nexus();
+        bytes memory callData = abi.encodeWithSelector(Nexus.upgradeToAndCall.selector, address(newSmartAccount), "");
+
+        Execution[] memory execution = new Execution[](1);
+        execution[0] = Execution(address(BOB_ACCOUNT), 0, callData);
+
+        PackedUserOperation[] memory userOps = buildPackedUserOperation(BOB, BOB_ACCOUNT, EXECTYPE_DEFAULT, execution, address(VALIDATOR_MODULE));
+        ENTRYPOINT.handleOps(userOps, payable(address(BOB.addr)));
+
+        address newImplementation = BOB_ACCOUNT.getImplementation();
```

Comment:
you mean this is failing currently?

Reply:
No this one from v3 works perfectly, but on fork while upgrading a v2 like here it's failing that's why I have commented the test:

<img width="1103" alt="Screenshot 2024-05-20 at 7 11 43 PM" src="https://github.com/bcnmy/nexus/assets/22845388/bb9ea983-c19f-42cc-b21e-f019039d1d3a">

Reply:
yes yes I meant after the upgrade. 
perhaps we have to remove getImplementation from the account 
larger impact to investigate would be by testing 
v2 -> V3 -> V4(some other implementation) 
cause 
storage slot for V2 is address of self
storage slot for V3 is 1967_SLOT